backend/certificates.js

```diff
@@ -0,0 +1,173 @@
+    }).catch((e) => {
+        Logger.log(`CA Server fails to responsd. ${JSON.stringify(e)}`);
+        if (e.code === 'ECONNREFUSED') {
+            Logger.log('Caught ECONNREFUSED from CA server.');
+            // If server is down, increase TIMEOUT to try a 1 minute later.
+            setCertificatesSettings('TIMEOUT_BETWEEN_TRIES', 60);
+        }
+    });
+}
+
+export function mockFetchCertsFromCA() {
+    return new Promise((resolve, reject) => {
+        resolve(fakeCerts);
+    });
+}
+
+export function fetchAndSaveCerts() {
+    /*
+     * This returns a promise to save certificates.
+     * When developing and debugging, you may want to quickly use the
+     * mocked certs (MOCK_CERTS=true) returned as a promise.
+     */
+    let fetchCerts;
+    if (LOCAL_SETTINGS.USE_MOCK_CERTS) {
+        fetchCerts = mockFetchCertsFromCA;
+    } else {
+        fetchCerts = fetchCertsFromCA;
+    }
+    return fetchCerts().then(response => {
+        if (!response.key || !response.cert || !response.subdomain) {
+            throw 'CA did not return one or more of [key, cert, subdomain].';
+        }
+        return saveCertsLocally(response);
+    });
+}
+
+// Wrapper around fetchAndSaveCerts to try again with a sleep period specified.
+
+export function timeoutFetchAndSaveCerts() {
+    // Increment tries.
+    setCertificatesSettings('ONGOING_COUNT', LOCAL_SETTINGS.ONGOING_COUNT + 1);
+
+    fetchAndSaveCerts()
+    .then (() => {
+        Logger.log('Successfully received certs from CA.');
+    }).catch((e) => {
+        Logger.log(`Failed to receive certs from CA. ${e}`);
+        // Retry calling CA if less than max count of tries.
+        if (LOCAL_SETTINGS.ONGOING_COUNT < LOCAL_SETTINGS.MAX_TRIES_COUNT) {
+            setTimeout(() => {
+                timeoutFetchAndSaveCerts();
+            }, LOCAL_SETTINGS.TIMEOUT_BETWEEN_TRIES * 1000);
+        }
+    });
+}
+
+export const msInOneDay = 1000 * 3600 * 24;
+
+export function calculateDaysToRenewal() {
+    // setInterval is using a 32 bit int to store the delay so the max value allowed
+    // would be 2147483647 => 24.8 days.
+    const {lastUpdated} = getSetting('CONNECTOR_HOST_INFO');
+    const timePassed = new Date().getTime() - new Date(lastUpdated).getTime();
+    return Math.ceil(daysCertificateIsGoodFor - timePassed / msInOneDay);
+}
+
+export function toMilliseconds(days) {
+    return days * msInOneDay;
+}
+
+// Renewing is the same as fetching a new cert because we are
+// generating a new hash each time.
+export const renewCertificate = () => {
+    fetchAndSaveCerts().then(() => {
+        setRenewalJob();
+    })
+    .catch((e) => {
+        Logger.log('Error occured during rewnewal of the certificate' + JSON.stringify(e));
+        Logger.log('Trying again.');
+        setTimeout(renewCertificate, 5000);
+    });
+};
+
+export function setRenewalJob() {
+    return setTimeout(() => {
+        renewCertificate();
+    }, toMilliseconds(calculateDaysToRenewal()));
```

Comment:
what if it's more than 30 days old? does setTimeout have the same 32 bit limitation? If so, we could just update this to `Math.max(toMilliseconds(calculateDaysToRenewal()), 0)`